### PR TITLE
Make returned participant errors during replication discoverable by the coordinator

### PR DESCRIPTION
**Pull request by redouan-rhazouani**
### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/replication.go [hunk 1]
```diff
@@ -46,186 +46,182 @@ type Replicator interface {
...
+	if localShard.isReadOnly() {
+		return nil, &replica.SimpleResponse{Errors: []replica.Error{{
+			Code: replica.StatusReadOnly, Msg: name,
+		}}}
+	}
+	return localShard, nil
+}
+
+func (i *Index) ReplicateObject(ctx context.Context, shard, requestID string, object *storobj.Object) replica.SimpleResponse {
+	localShard, pr := i.writableShard(shard)
```
Comment (parkerduckworth) [inline adapters/repos/db/replication.go:162]: Why is this `SimpleResponse` variable called `pr`?

File: adapters/repos/db/shard_replication.go [hunk 2]
```diff
@@ -117,15 +132,12 @@ func (s *Shard) prepareDeleteObject(ctx context.Context, requestID string, uuid
...
-		return replica.SimpleResponse{Errors: []string{storagestate.ErrStatusReadOnly.Error()}}
-	}
 	task := func(ctx context.Context) interface{} {
 		rawErrs := s.putBatch(ctx, objects)
-		resp := replica.SimpleResponse{Errors: make([]string, len(rawErrs))}
+		resp := replica.SimpleResponse{Errors: make([]replica.Error, len(rawErrs))}
 		for i, err := range rawErrs {
 			if err != nil {
-				resp.Errors[i] = err.Error()
+				resp.Errors[i] = replica.Error{Code: replica.StatusConflict, Msg: err.Error()}
```
Comment (parkerduckworth) [inline adapters/repos/db/shard_replication.go:140]: Can we be sure the error scenario will always be a conflict here?

File: adapters/repos/db/shard_replication.go [hunk 3]
```diff
@@ -147,7 +156,7 @@ func (s *Shard) prepareDeleteObjects(ctx context.Context, requestID string, docI
 		for i, r := range result {
 			entry := replica.UUID2Error{UUID: string(r.UUID)}
 			if err := r.Err; err != nil {
-				entry.Error = err.Error()
+				entry.Error = replica.Error{Code: replica.StatusConflict, Msg: err.Error()}
```
Comment (parkerduckworth) [inline adapters/repos/db/shard_replication.go:159]: I think I am not clear on what `conflict` is indicating here

File: usecases/replica/transport.go [hunk 4]
```diff
@@ -25,14 +25,85 @@ const (
...
+
+const (
+	StatusOK            = 0
+	StatusClassNotFound = iota + 200
+	StatusShardNotFound
+	StatusNotFound
+	StatusAlreadyExisted
+	StatusConflict = iota + 300
+	StatusPreconditionFailed
+	StatusReadOnly
```
Comment (parkerduckworth) [inline usecases/replica/transport.go:38]: What is the basis for the values of these status codes?

File: adapters/repos/db/replication.go (hunk 1 above)
Comment (redouan-rhazouani) [reply adapters/repos/db/replication.go:162]: pr is a local variable which stands for pointer to response. pr == nil  if localshard exists

File: adapters/repos/db/shard_replication.go (hunk 2 above)
Comment (redouan-rhazouani) [reply adapters/repos/db/shard_replication.go:140]: Right now everything which prevents the commit()  from being executed is marked as conflict.

File: adapters/repos/db/shard_replication.go (hunk 3 above)
Comment (redouan-rhazouani) [reply adapters/repos/db/shard_replication.go:159]: Same as before

**Review by parkerduckworth — APPROVED** [review]

**Pull request merged**